Alex_RTOS_Sailboat/RTOS_framework_v1/src/sail_gps.c

```diff
@@ -0,0 +1,344 @@
+/*
+ * sail_gps.c
+ *
+ * Created: 2022-06-02 8:04 PM
+ *  Author: secul
+ */
+
+#include "sail_gps.h"
+#include <stdio.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <string.h>
+#include <ctype.h>
+#include <stdlib.h>
+#include "FreeRTOS.h"
+#include "FreeRTOSConfig.h"
+#include "task.h"
+#include "sail_ctrl.h"
+#include "sail_tasksinit.h"
+#include "usart_interrupt.h"
+
+#include "sail_debug.h"
+
+#define HEADER_FMT				"DALSAIL,%03"PRIu16
+#define FIELD_FMT				",%"PRIi32
+
+#define RADIO_BUFFER_LENGTH		NMEA_BUFFER_LENGTH
+
+static bool init_flag = false;
+static char msg_buffer[RADIO_BUFFER_LENGTH];
+
+//stores all the msg types of the gps sensor
+volatile GPS_AllMsgs GPS_data;
+
+void ReadGPS(void) {
+	DEBUG_Write("Reading GPS...\r\n");
+	uint16_t loop_cnt = 0;
+	//set msg type sum to 0 since no messages processed yet
+	GPS_data.msg_type_sum = 0;
+
+	// Event bits for holding the state of the event group
+	EventBits_t event_bits;
+
+	TickType_t read_gps_delay = pdMS_TO_TICKS(GPS_SLEEP_PERIOD_MS);
+
+	while (1) {
+
+		event_bits = xEventGroupWaitBits(mode_event_group,                        /* Test the mode event group */
+			CTRL_MODE_AUTO_BIT | CTRL_MODE_REMOTE_BIT, /* Wait until the sailboat is in AUTO or REMOTE mode */
+			pdFALSE,                                 /* Bits should not be cleared before returning. */
+			pdFALSE,                                 /* Don't wait for both bits, either bit will do. */
+			portMAX_DELAY);                          /* Wait time does not expire */
+			
+		taskENTER_CRITICAL();
+		watchdog_counter |= 0x01;
+		taskEXIT_CRITICAL();
+
+		GPS_On();
```

Comment:
Should we toggle a flag to know if we are on instead of calling this every time?

Reply:
The flag is init_flag is being set in GPS_Init(). GPS_On() is calling GPS_Init() and GPS_Enable(). Basically, those methods are there to enable the gps as long as the NMEA channel is working/initialize. If there is something wrong with the NMEA channel, then the GPS will not be initialized (init_flag remains set to False). 

Once GPS is initialized, the GPS_init() will return with status OK. However, GPS_Enable may not return status OK as it checks the NMEA channel. Might be some levels of redundancy that are not needed here... Should probably go over this and make sure the error calling is being handled properly